### PR TITLE
Partial modernization of docs

### PR DESCRIPTION
**Pull request by termi-official**
This PR aims to improve the docs in two ways. 
1. The standard doxygen design is replaced with a more modern design (https://jothepro.github.io/doxygen-awesome-css/index.html).
2. I try to improve the doc strings for all integrators and error estimators by using latex instead of ascii. Unicode docs are mostly left untouched for now.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3688](https://github.com/mfem/mfem/pull/3688) | @termi-official | @tzanio | @acfisher + @tzanio + @mlstowell + @v-dobrev | 11/29/23 | 12/13/23 | 12/17/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by termi-official** [comment]
Is this something we might also want to include in the next release (4.6) or does it require a longer discussion @tzanio @v-dobrev ?

**Comment by acfisher** [comment]
I'll see if I can get the ball rolling on pulling this in.

**Comment by acfisher** [comment]
I just built pulled the branch and built it, and it looks like a significant improvement to the look and feel.

**Comment by acfisher** [comment]
@kmittal2 @psocratis would you be available to review this in the near future?

**Comment by termi-official** [comment]
If things should be done differently feel free to note this here. I think we can work on guidelines for doc strings containing math alongside this PR and update the contributor guidelines to ensure a consistent experience in the docs, as well as a starting point for contributors. But we can also make this a separate issue and PR.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
I like the new Doxygen style in this PR, thanks @termi-official !

Here are some comparisons: 

| old style | new style |
| ------------- | ------------- |
|  Screenshot 2023-11-29 at 4 42 21 PM |  Screenshot 2023-11-29 at 4 42 24 PM |
| Screenshot 2023-11-29 at 4 45 16 PM | Screenshot 2023-11-29 at 4 45 18 PM |
| Screenshot 2023-11-29 at 4 51 31 PM | Screenshot 2023-11-29 at 4 51 33 PM |
| Screenshot 2023-11-29 at 4 53 29 PM | Screenshot 2023-11-29 at 4 53 31 PM |

**Review by tzanio — APPROVED** [review]
😎

**Comment by termi-official** [comment]
Happy to see that it is not only me liking the style!

Regarding the doc string guidelines, my idea was that we try to enforce tex math and give some examples + the doxygen reference. Regarding unicode math strings, like for example in the doc string for the KellyErrorEstimator, I am not sure if there should be a rule. It gives a slight inconsistens look and feel, but unicode math strings have the advantage that they are easier to read in the source.

**Comment by pazner** [comment]
I would suggest making the following small change
```diff
diff --git a/doc/makefile b/doc/makefile
index 4012074dd..738e223ec 100644
--- a/doc/makefile
+++ b/doc/makefile
@@ -16,7 +16,7 @@ DOXYGEN_CONF = CodeDocumentation.conf
 # doxygen uses: graphviz, latex
 html: $(DOXYGEN_CONF)
 	@# Generate the html documentation
-	@( cat $(DOXYGEN_CONF) ; echo "$(MFEM_DOXYGEN_FLAGS)" ) | doxygen -
+	@( cat $(DOXYGEN_CONF) ; printf "$(MFEM_DOXYGEN_FLAGS)\n" ) | doxygen -
 	@echo "<meta http-equiv=\"REFRESH\" content=\"0;URL=CodeDocumentation/html/index.html\">" > CodeDocumentation.html
 	@cat warnings.log 1>&2
 	@# Generate the log of undocumented methods

```
which would allow multiple command-line configurations like
```
make MFEM_DOXYGEN_FLAGS="HAVE_DOT = NO\nHTML_EXTRA_STYLESHEET = "
```
to make it easy for someone to switch back to the default theme (or to a different theme).

**Comment by tzanio** [comment]
@termi-official -- Do you know what are the customization options for Doxygen Awesome? Can we for example switch dark mode on/off with a button or have the left panel as in https://jothepro.github.io/doxygen-awesome-css/todo.html?

**Comment by termi-official** [comment]
> @termi-official -- Do you know what are the customization options for Doxygen Awesome? Can we for example switch dark mode on/off with a button or have the left panel as in https://jothepro.github.io/doxygen-awesome-css/todo.html?

Sorry @tzanio I am not sure if I understand the first question correctly, but I think you are asking for what can be put into the "custom.css" file. In this file you can basically just overwrite all existing css options made by Doxygen Awesome (and hence Doxygen) found here https://github.com/jothepro/doxygen-awesome-css/blob/df88fe4fdd97714fadfd3ef17de0b4401f804052/doxygen-awesome.css#L30. The docs page is here https://jothepro.github.io/doxygen-awesome-css/md_docs_2customization.html and some examples can be found here https://github.com/jothepro/doxygen-awesome-css/discussions/13 .

Regarding the dark mode toggle, see my last commit for support. Some extensions and installation instructions are described here https://github.com/jothepro/doxygen-awesome-css/blob/main/docs/extensions.md . Does this answer the latter question?

File: doc/html/header.html
Comment (tzanio) [inline doc/html/header.html:22]: This was showing up for me as a `$darkmode` text on the top of the page.

Things seem fine without it... can we remove it?

Comment (tzanio) [reply doc/html/header.html:22]: OK, I upgraded to 1.9.8 and this was fixed, but I had to change something else to get the light mode to work properly on my system (Mac with "dark mode" by default) -- I was getting a grey/black instead of light/dark.

Specifically, I had to set `HTML_COLORSTYLE = LIGHT`

As a part of this change I took the opportunity to upgrade `CodeDocumentation.conf.in`. Do you want me to create a branch for those changes, or should I just push here.

Comment (termi-official) [reply doc/html/header.html:22]: Great. Feel free to push into the branch. Which version did you have before? We could throw a warning if the Doxygen version is too old.

**Comment by termi-official** [comment]
Works locally for me. I think we should leave it to *Default* for now and can introduce a sidebar later, if there are plans for a more involved docs structure which beneifts from it. Without the sidebar the docs also fit better on half of a screen (if you work with multiple parallel windows per screen).

**Comment by tzanio** [comment]
@v-dobrev, @jandrej and @pazner -- do you think we should go forward with this?

**Review by pazner — APPROVED** [review]
Looks good to me, and it's not difficult to locally go back to the old style if someone prefers. The LaTeX documentation is great. Thanks @termi-official!

**Comment by v-dobrev** [comment]
We may need to update the `doxygen` versions we use in CI and for the nightly tests. Is v1.9.8 the minimal version we need (as specified in the new `CodeDocumentation.conf.in`)? The github action uses v1.9.1 and ends up ignoring and removing a number of options from the new config file, e.g. see https://github.com/mfem/mfem/actions/runs/7133157275/job/19425233629?pr=3688#step:4:29.

**Comment by termi-official** [comment]
With the current configuration, yes, 1.9.8 is the compat bound (see https://www.doxygen.nl/manual/changelog.html#log_1_9_8). If that does not work then we can also tweak the config to some older version. However, I would prefer to update doxygen here if possible, since we already upgrade the docs, also because 1.9.1 is already almost 3 years old and includes a good chunk of bugfixes.

**Comment by tzanio** [comment]
Is [this](https://github.com/mfem/mfem/blob/master/.github/workflows/repo-check.yml#L108) the only place we need to change the Doxygen version for the CI? 

Is there an easy way to see what Doxygen packages are available in the GH image we are using for testing?

**Comment by v-dobrev** [comment]
The official Ubuntu packages do not have newer version of doxygen for jammy (22.04 LTS), see https://packages.ubuntu.com/search?keywords=doxygen.

**Comment by tzanio** [comment]
@v-dobrev, the doxygen version requirement is now mentioned in https://github.com/mfem/mfem/pull/3688/commits/a8b0b35c1d58a5dbacc821b2880d6816299790cd

**Comment by tzanio** [comment]
Can I get more approvals, please?

**Review by acfisher — APPROVED** [review]
There is are some great improvements here.  I'm glad to see these getting pulled in.

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: doc/CodeDocumentation.conf.in
Comment (pazner) [inline doc/CodeDocumentation.conf.in:202]: While we're thinking about Doxygen, do we want to enable [autobrief](https://www.doxygen.nl/manual/config.html#cfg_javadoc_autobrief)? This means we won't have to write `@brief` all the time (and has few downsides as far as I can tell).

Comment (tzanio) [reply doc/CodeDocumentation.conf.in:202]: I like that suggestion 👍

Comment (termi-official) [reply doc/CodeDocumentation.conf.in:202]: While I am not opposed to the idea I would like to leave this to a separate PR as this is not so trivial to change. Take the docs for `NodesUpdated` as an example, where "e.g." is used, which is interpreted as  the end of the brief part.

**Comment by termi-official** [comment]
I see that the CI for the merge failed, but I cannot see any explicit error. Is this some Github server hiccup or am I missing something else?

File: doc/CodeDocumentation.conf.in
Comment (pazner) [reply doc/CodeDocumentation.conf.in:202]: OK, let's keep it as a separate PR.

I assumed that you could "override" the autobrief docs with an explicit `@brief`, but it appears that doesn't work.

Another (better?) option would be to enable `MULTILINE_CPP_IS_BRIEF`. This would make the first paragraph of a `///` comment block the brief comment, and subsequent paragraphs would be the detailed description.

**Comment by pazner** [comment]
> I see that the CI for the merge failed, but I cannot see any explicit error. Is this some Github server hiccup or am I missing something else?

I think the CI for the merge commit into `next` was just canceled because there was another commit to `next` before all the jobs finished. Everything looks OK to me.

**Comment by tzanio** [comment]
Thanks for suggesting this refresh, @termi-official!

Let's keep pushing on the documentation to make it better 👍

**Pull request merged**
